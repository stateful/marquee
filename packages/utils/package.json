{
  "name": "@vscode-marquee/utils",
  "description": "Marquee utils",
  "version": "0.1.0",
  "private": true,
  "main": "./build",
  "types": "./build/index.d.ts",
  "dependencies": {
    "@types/lodash.pick": "^4.4.6",
    "@types/object-hash": "^2.2.1",
    "@types/uuid": "^8.3.4",
    "hex-rgb": "4.3.0",
    "lodash.pick": "^4.4.0",
    "object-hash": "^3.0.0",
    "uuid": "^8.3.2"
  },
  "devDependencies": {},
  "peerDependencies": {
    "@types/react": "^17.0.39",
    "@types/vscode": "^1.63.0",
    "react": "^17.0.2"
  },
  "peerDependenciesMeta": {
    "@types/react": {
      "optional": true
    },
    "@types/vscode": {
      "optional": true
    }
  }
}
