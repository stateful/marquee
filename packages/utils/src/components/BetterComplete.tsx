import React, { ChangeEvent } from 'react'
import Autocomplete from '@mui/material/Autocomplete'
import { styled, TextField } from '@mui/material'

export interface Props {
  id: string
  field: string,
  options: any,
  display: string
  label: string
  variant: 'filled' | 'standard' | 'outlined' | undefined
  value: any
  isOptionEqualToValue: (option: any, value: any) => boolean
  onChange: (e: ChangeEvent<{}>, v: any) => void
}

const PREFIX = 'BetterComplete'
const classes = {
  option: `${PREFIX}-option`,
  input: `${PREFIX}-input`,
  label: `${PREFIX}-label`,
}

const Root = styled('div')(() => ({
  [`&.${classes.option}`]: {
    backgroundColor: 'var(--vscode-editor-background)',

    '&[data-focus="true"]': {
      backgroundColor: 'var(--vscode-editor-selectionBackground)',
      color: 'var(--vscode-input-foreground)',
    },
    '&[aria-selected="true"]': {
      backgroundColor: 'var(--vscode-editor-selectionBackground)',
      color: 'var(--vscode-editor-foreground)',
    },
  },
  [`& .${classes.input}`]: {
    // color: "black",
    color: 'var(--vscode-editor-foreground)',
  },
  [`& .${classes.label}`]: {
    color: 'var(--vscode-editor-foreground)',
  },
}))

const BetterComplete = ({
  id,
  field,
  options,
  display,
  label,
  variant,
  ...rest
}: Props) => {

  return (
    <Root>
      <Autocomplete
        id={id}
        classes={{
          option: classes.option,
          input: classes.input,
        }}
        options={options}
        getOptionLabel={(option) => option[display] || ''}
        renderInput={(params) => (
          <TextField
            {...params}
            label={label || field}
            fullWidth
            variant={variant || 'outlined'}
            InputLabelProps={{
              className: classes.label,
            }}
          />
        )}
        {...rest}
      />
    </Root>
  )
}

export default React.memo(BetterComplete)
