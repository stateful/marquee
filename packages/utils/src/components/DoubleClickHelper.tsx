import React from 'react'
import { IconButton, Tooltip } from '@mui/material'

import HelpIcon from '@mui/icons-material/Help'

export interface DoubleClickHelperParams {
  content?: string
}

let DoubleClickHelper = (props: DoubleClickHelperParams) => {
  const content = props.content || 'Double-click item titles to edit'
  return (
    <Tooltip title={content}>
      <IconButton size="small">
        <HelpIcon fontSize="small" />
      </IconButton>
    </Tooltip>
  )
}

export default React.memo(DoubleClickHelper)
