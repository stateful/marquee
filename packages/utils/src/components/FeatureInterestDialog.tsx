import React, { useState } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Button, DialogActions, DialogContent, Link, Typography } from '@mui/material'
import { DialogContainer, DialogTitle } from '@vscode-marquee/dialog'
import { faDiscord } from '@fortawesome/free-brands-svg-icons/faDiscord'

const discordLink = 'https://discord.gg/BQm8zRCBUY'

interface FeatureInterestDialogProps {
  setShowCloudSyncFeature: (show: boolean) => void
  _isInterestedInSyncFeature: (interested: boolean) => void
}
const FeatureInterestDialog = ({ setShowCloudSyncFeature, _isInterestedInSyncFeature }: FeatureInterestDialogProps) => {
  const [showCloudSyncFeedbackResponse, setShowCloudSyncFeedbackResponse] = useState(false)

  const DialogHeader = showCloudSyncFeedbackResponse
    ? 'Thank you for your response.'
    : 'Do you want to sync your Notes, Todos and Snippets across multiple machines?'

  const DialogBody = showCloudSyncFeedbackResponse
    ? <DialogContent>
      Join our
      <Link
        component="a"
        href={discordLink}
        target="_blank"
        underline="hover">
        &nbsp;discord &nbsp;
        <FontAwesomeIcon icon={faDiscord} />
      </Link>
      {' '}channel if you like to stay updated with
      upcoming features or if you have other feature requests.
    </DialogContent>
    : <DialogActions>
      <Button onClick={() => {
        _isInterestedInSyncFeature(false)
        setShowCloudSyncFeedbackResponse(true)
      }}>No</Button>
      <Button onClick={() => {
        _isInterestedInSyncFeature(true)
        setShowCloudSyncFeedbackResponse(true)
      }}>
        Yes
      </Button>
    </DialogActions>

  return (
    <DialogContainer fullWidth={true} onClose={() => setShowCloudSyncFeature(false)} >
      <DialogTitle onClose={() => setShowCloudSyncFeature(false)} >
        <Typography style={{ width: '85%', fontWeight: 'bold' }}>
          {DialogHeader}
        </Typography>
      </DialogTitle>
      {DialogBody}
    </DialogContainer>
  )
}

export default React.memo(FeatureInterestDialog)