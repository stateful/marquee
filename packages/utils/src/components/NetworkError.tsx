import React from "react";
import { IconButton, Link, Grid, Typography } from "@mui/material";
import ErrorIcon from "@mui/icons-material/Error";
import type { MarqueeWindow } from '../types';

declare const window: MarqueeWindow;

export interface NetworkErrorProps {
  message?: string | JSX.Element
}

const NetworkError = ({ message }: NetworkErrorProps) => {
  const errorCaption = message || (
    <>
      Are you using a proxy? Please make sure to set it up in the {" "}
      <Link
        style={{ cursor: 'pointer' }}
        onClick={() => window.vscode.postMessage({
          west: { execCommands: [{
            command: 'workbench.action.openSettings',
            args: ['Marquee']
          }],
        }})}
        underline="hover">
          Marquee settings
      </Link>.
    </>
  );

  return (
    <Grid
        container
        style={{ height: "100%" }}
        alignItems="center"
        justifyContent="center"
        direction="column"
    >
        <Grid item>
        <Typography variant="subtitle1">
            <IconButton size="small">
            <ErrorIcon fontSize="small" />
            </IconButton>
            <span>Error fetching data!</span>
        </Typography>
        <Typography variant="caption">
          {errorCaption}
        </Typography>
        </Grid>
    </Grid>
  );
};

export default React.memo(NetworkError);
