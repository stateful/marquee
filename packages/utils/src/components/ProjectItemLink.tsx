import React, { useMemo } from 'react'
import LinkIcon from '@mui/icons-material/Link'
import Tooltip from '@mui/material/Tooltip'
import { Link, Button, Typography, IconButton } from '@mui/material'

import { jumpTo } from '../index'
import type { ProjectItem, MarqueeWindow } from '../types'

const marqueeWindow: MarqueeWindow = window as any

const transformPathToLink = (item: ProjectItem) => {
  try {
    // transform
    // -> git@github.com/foo/bar.git#main
    // to
    // -> https://github.com/foo/bar/blob/adcbe2a5c0428783fe9d6b50a1d2e39cbbe2def6/some/file#L3
    const [path, line] = item.origin!.split(':')
    const u = new window.URL(`https://${item.gitUri!
      .replace(':', '/')
      .split('@')
      .pop()!
    }`)
    const rPath = path.replace(marqueeWindow.activeWorkspace!.path, '')
    return `${u.origin}${u.pathname.replace(/\.git$/, '')}/blob/${item.commit}${rPath}#L${parseInt(line, 10) + 1}`
  } catch (err: any) {
    console.warn(`Couldn't construct remote url: ${(err as Error).message}`)
    return undefined
  }
}

interface ProjectItemLinkParams {
  item?: ProjectItem
  iconOnly?: boolean
}

let ProjectItemLink = (props: ProjectItemLinkParams) => {
  if (!props.item) {
    return <></>
  }

  const itemLinkName = useMemo(() => {
    const path = props.item?.path || props.item?.origin || 'unknown:unknown'
    const [fileName, line] = path.split('/').pop()!.split(':')
    return `${fileName}:${parseInt(line, 10) + 1}`
  }, [props.item])

  const link = useMemo(() => {
    if (props.item?.path) {
      return props.item?.path
    }
    const useRemoteLink = props.item?.gitUri && props.item?.commit && props.item?.origin
    if (useRemoteLink && marqueeWindow.activeWorkspace) {
      return transformPathToLink(props.item!)
    }
  }, [props.item])

  if (props.item.path) {
    if (props.iconOnly) {
      return (
        <Tooltip
          style={{borderRadius: '4px'}}
          title={<Typography variant="subtitle2">{props.item.path}</Typography>}
          placement="top"
          arrow
        >
          <Typography variant="body2" noWrap>
            <IconButton aria-label="Project Item Link" size="small" tabIndex={-1} onClick={() => jumpTo(props.item)}>
              <LinkIcon />
            </IconButton>
          </Typography>
        </Tooltip>
      )
    }

    return (
      <Button
        aria-label="Project Item Link"
        size="small"
        startIcon={<LinkIcon />}
        disableFocusRipple
        onClick={() => jumpTo(props.item)}
        style={{
          padding: '0 5px',
          background: 'transparent',
          color: 'inherit'
        }}
      >
        {itemLinkName}
      </Button>
    )
  }

  if (!link) {
    return <></>
  }

  if (props.iconOnly) {
    return (
      <Tooltip
        style={{borderRadius: '4px', marginLeft: '5px'}}
        title={<Typography variant="subtitle2">{link}</Typography>}
        placement="top"
        arrow
      >
        <Typography variant="body2" noWrap>
          <Link aria-label="Project Item Link" tabIndex={-1} href={link}>
            <LinkIcon />
          </Link>
        </Typography>
      </Tooltip>
    )
  }

  return (
    <Link
      aria-label="Project Item Link"
      href={link}
      style={{
        padding: '3px 5px 0 29px',
        background: 'transparent',
        color: 'inherit',
        fontSize: '.81em',
        textDecoration: 'none',
        position: 'relative',
        display: 'block'
      }}
    >
      <LinkIcon style={{
        fontSize: '18px',
        position: 'absolute',
        top: '2px',
        left: '3px'
      }} />
      {itemLinkName}
    </Link>
  )
}

export default React.memo(ProjectItemLink)
