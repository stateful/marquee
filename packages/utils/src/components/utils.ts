import type { MarqueeWindow } from '../types'

declare const window: MarqueeWindow

/**
 * open the reference of a project item
 * @param item project item (note/snippet/todo)
 * @param w    window object (for testing purposes only)
 */
export const jumpTo = (item: any, w = window) => {
  w.vscode.postMessage({
    west: {
      execCommands: [
        {
          command: 'marquee.link',
          args: [{ item }],
        },
      ],
    },
  })
}
