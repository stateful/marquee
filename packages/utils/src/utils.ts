import hexRgb from 'hex-rgb'

import type { RGBA } from '../src/types'

export const getVSColor = () => {
  let vsCodeBgColor = process.env.NODE_ENV === 'test'
    ? '#000000'
    : getComputedStyle(document.documentElement)
      .getPropertyValue('--vscode-sideBar-background')

  let vsRgba = hexRgb(vsCodeBgColor)
  return {
    r: vsRgba.red,
    g: vsRgba.green,
    b: vsRgba.blue,
    a: 0.8,
  } as RGBA
}
