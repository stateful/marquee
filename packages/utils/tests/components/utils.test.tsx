import { jumpTo } from '../../src/components/utils'

test('jumpTo', () => {
  const window = { vscode: { postMessage: jest.fn() } }
  jumpTo({ foo: 'bar' }, window as any)
  expect(window.vscode.postMessage.mock.calls).toMatchSnapshot()
})
