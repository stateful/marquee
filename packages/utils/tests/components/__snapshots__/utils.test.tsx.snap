// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jumpTo 1`] = `
Array [
  Array [
    Object {
      "west": Object {
        "execCommands": Array [
          Object {
            "args": Array [
              Object {
                "item": Object {
                  "foo": "bar",
                },
              },
            ],
            "command": "marquee.link",
          },
        ],
      },
    },
  ],
]
`;
